Add and use meson-multilib.eclass

mattst88 opened the pull request:
There are a lot of private (i.e. to individual ebuilds and/or packages) functions implementing the necessary `meson_use`/`meson_feature`-style functions for multilib.

This PR factors those out and shows how we can use them in existing packages.

Consider this an RFC. I'm open to all feedback and I don't feel strongly about anything I've written here.

A couple of unorganized thoughts:

- I feel like the function names are too long. Should we remove `multilib_` from the names?

- Trying to make the function names match `meson_use`/`meson_feature` exposes the fact that the `meson_use` name is weird. We want functions for the different meson options:
    - `feature` (enabled/disabled)
    - `bool` (true/false)
    - maybe something for yes/no (a common `combo` configuration).

Within those, we want multilib functions that operate with and without a USE flag. One natural way to name those functions differently is to put `_use` in the name of the ones that take a USE flag, e.g., could be something like `meson_multilib_native_use_enable` vs `meson_multilib_native_enable`. But if we already are using `_use` in a name like `meson_multilib_native_use` then adding another `_use` is kind of nonsensical. It's for that reason that I instead have `meson_multilib_native_use` and `meson_multilib_native_true`. Perhaps we should consider deprecating `meson_use` and replacing it with a `meson_bool`? :shrug:

floppym commented:
> I feel like the function names are too long. Should we remove multilib_ from the names?

Yeah, I think "native" already conveys the multilib-nature of the functions.

> Trying to make the function names match meson_use/meson_feature exposes the fact that the meson_use name is weird.

I like the idea of including the output type in the function name.

mattst88 commented:
Removed `_multilib`  from the names and changed `meson_native_use` -> `meson_native_bool` to better indicate the return type.

Also added a patch to `meson.eclass` to cd into `${S}` around `einstalldocs` like cmake.eclass does.

Still need eclass documentation for `meson-multilib.eclass`.

mattst88 commented:
meson-multilib.eclass documentation completed. I've sent the two eclass patches at the beginning of the series to gentoo-dev@ for review.

SoapGentoo commented on eclass/meson-multilib.eclass:
```diff
@@ -0,0 +1,126 @@
+inherit meson multilib-minimal
+
+EXPORT_FUNCTIONS src_configure src_compile src_test src_install
+
```
missing inherit guard

mattst88 replied:
Is it needed? `cmake-multilib.eclass` doesn't have it, and both `meson.eclass` and `multilib-minimal.eclass` have it?

Doesn't hurt to have it I guess.

SoapGentoo commented on media-libs/harfbuzz/harfbuzz-9999.ebuild:
```diff
@@ -100,14 +92,6 @@ multilib_src_configure() {
-	meson_src_install
-}
-
 multilib_src_install_all() {
```
is this one necessary?

mattst88 replied:
Nope. Thanks!

SoapGentoo commented on media-libs/mesa/mesa-20.3.5.ebuild:
```diff
@@ -526,14 +526,6 @@ multilib_src_configure() {
-	meson_src_install
-}
-
 multilib_src_install_all() {
```
necessary?

mattst88 replied:
Nope!

SoapGentoo commented on media-libs/mesa/mesa-21.0.3.ebuild:
```diff
@@ -524,14 +524,6 @@ multilib_src_configure() {
-	meson_src_install
-}
-
 multilib_src_install_all() {
```
necessary?

mattst88 replied:
Nope!

SoapGentoo commented on media-libs/mesa/mesa-21.1.1.ebuild:
```diff
@@ -525,14 +525,6 @@ multilib_src_configure() {
-	meson_src_install
-}
-
 multilib_src_install_all() {
```
necessary?

mattst88 replied:
Nope!

SoapGentoo commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -524,14 +524,6 @@ multilib_src_configure() {
-	meson_src_install
-}
-
 multilib_src_install_all() {
```
necessary?

mattst88 replied:
Nope!

SoapGentoo commented on x11-libs/libxkbcommon/libxkbcommon-1.3.0.ebuild:
```diff
@@ -55,15 +55,3 @@ multilib_src_configure() {
-}
-
-multilib_src_test() {
-	virtx meson_src_test
```
sure?

mattst88 replied:
Oops. I didn't catch this because the only one of the tests needs an Xserver and it nicely skips if it doesn't have one.

mattst88 commented:
Found and replaced a few more `-Doption=$(usex ... true false)` and removed some more unnecessary einstalldocs.

Planning to push this tomorrow if there are no further comments.

mgorny commented on eclass/meson-multilib.eclass:
```diff
@@ -0,0 +1,131 @@
+
+# @ECLASS: meson-multilib.eclass
+# @MAINTAINER:
+# gx86-multilib team <multilib@gentoo.org>
```
ENOSUCHTEAM.

mattst88 replied:
You're so sneaky!

mgorny commented on eclass/meson-multilib.eclass:
```diff
@@ -0,0 +1,131 @@
+# if building for the native ABI (multilib_is_native_abi is true). Otherwise,
+# output -Doption=false. If the project option is unspecified, it defaults
+# to the USE flag.
+meson_native_bool() {
```
Personally, I find the naming to be confusing but I get your point. Still, I'd prefer if stuff using USE flags had `use` in its name.

mattst88 replied:
What should the names be then? How about:

`meson.eclass`:
* `meson_use` → `meson_use_bool`
* `meson_feature` → `meson_use_feature`

`meson-multilib.eclass`:
* `meson_native_bool` → `meson_native_use_bool`
* `meson_native_feature` → `meson_native_use_feature`
* `meson_native_true`
* `meson_native_enabled`

Seems to fit with `use_enable` and `use_with`.

mgorny replied:
I would go a step further and use either `bool` or `true` consistently, e.g. `meson_native_use_bool` and `meson_native_bool` etc.

mattst88 replied:
Discussed on IRC and mgorny said

> <mgorny> mattst88: at a second look, i'd probably use _true and _enabled

sam and floppym acked.

mgorny commented on app-crypt/libsecret/libsecret-0.20.4.ebuild:
```diff
@@ -71,16 +71,12 @@ src_prepare() {
 		# Don't build docs multiple times
-		-Dmanpage=$(multilib_is_native_abi && echo true || echo false)
+		$(meson_native_true manpage)
 		$(meson_multilib_native_use gtk-doc gtk_doc)
```
But you've renamed the function, right?

mattst88 replied:
Damn. Not sure how that slipped through. Fixed.
